### PR TITLE
Enable email notifier to send emails for health check failures

### PR DESCRIPTION
**Pull request by prakhar10**
This change enables to send email notifications for healthcheck failures on backend trino clusters.

New fields added in the `NotifierConfiguration.java` class:
`enabled` - To enable or disable sending emails for healthcheck failures
`gatewayInstance` - To specify gateway instance name in case there are multiple instances. This will be displayed in the email subject so that one can know in which instance the cluster is unhealthy.
`customContent` - To provide custom email body in case the default content is not very helpful.

### CODE REVIEW

**Comment by prakhar10** [comment]
> Thank you for your pull request and welcome to the Trino community. We require contributors to sign our [Contributor License Agreement](https://github.com/trinodb/cla/raw/master/Trino%20Foundation%20Individual%20CLA.pdf), and we don't seem to have you on file. Continue to work with us on the review and improvements in this PR, and submit the signed CLA to [cla@trino.io](mailto:cla@trino.io). Photos, scans, or digitally-signed PDF files are all suitable. Processing may take a few days. The CLA needs to be on file before we merge your changes. For more information, see https://github.com/trinodb/cla

I have already sent this.

File: gateway-ha/src/main/java/io/trino/gateway/ha/module/ClusterStateListenerModule.java [hunk 1]
```diff
@@ -52,8 +56,14 @@ public List<TrinoClusterStatsObserver> getClusterStatsObservers(
             BackendStateManager backendStateManager)
     {
         observers = new ArrayList<>();
+        NotifierConfiguration notifierConfiguration = getConfiguration().getNotifier();
+        Notifier emailNotifier = new EmailNotifier(notifierConfiguration);
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/module/ClusterStateListenerModule.java:60]: Please inline emailNotifier variable. No need to initialize when the notification is disabled.

**Review by ebyhr — COMMENTED** [review]
Just skimmed.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthChecker.java [hunk 2]
```diff
@@ -48,20 +48,33 @@ public void observe(List<ClusterStats> clustersStats)
 
     private void notifyUnhealthyCluster(ClusterStats clusterStats)
     {
-        notifier.sendNotification(String.format("%s - Cluster unhealthy",
-                        clusterStats.clusterId()),
-                clusterStats.toString());
+        notifier.sendNotification(String.format("Cluster name %s is unhealthy",
```
[the remark this answers is not in the document]
Comment (prakhar10) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthChecker.java]: done

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthChecker.java [hunk 3]
```diff
@@ -48,20 +48,33 @@ public void observe(List<ClusterStats> clustersStats)
...
+                        clusterStats.clusterId()), buildContent(clusterStats));
     }
 
     private void notifyForTooManyQueuedQueries(ClusterStats clusterStats)
     {
-        notifier.sendNotification(String.format("%s - Too many queued queries",
-                clusterStats.clusterId()), clusterStats.toString());
+        notifier.sendNotification(String.format("Cluster name %s has too many queued queries",
```
[the remark this answers is not in the document]
Comment (prakhar10) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthChecker.java]: done

File: gateway-ha/src/main/java/io/trino/gateway/ha/notifier/EmailNotifier.java [hunk 4]
```diff
@@ -54,10 +54,17 @@ public EmailNotifier(NotifierConfiguration notifierConfiguration)
...
+                    notifierConfiguration.getSender(),
+                    notifierConfiguration.getRecipients(),
+                    String.format("%s error: %s", notifierConfiguration.getGatewayInstance() == null ? "Trino" : notifierConfiguration.getGatewayInstance(), subject),
+                    notifierConfiguration.getCustomContent());
+        }
         sendNotification(
                 notifierConfiguration.getSender(),
                 notifierConfiguration.getRecipients(),
-                "Trino Error: " + subject,
+                String.format("%s error: %s", notifierConfiguration.getGatewayInstance() == null ? "Trino" : notifierConfiguration.getGatewayInstance(), subject),
```
[the remark this answers is not in the document]
Comment (prakhar10) [reply gateway-ha/src/main/java/io/trino/gateway/ha/notifier/EmailNotifier.java]: done

File: gateway-ha/src/main/java/io/trino/gateway/ha/module/ClusterStateListenerModule.java (hunk 1 above)
Comment (prakhar10) [reply gateway-ha/src/main/java/io/trino/gateway/ha/module/ClusterStateListenerModule.java]: done

**Comment by willmostly** [comment]
@prakhar10 could you squash your commits? 

I think this enhancement should include some tests, there are currently none for notifications.

File: gateway-ha/src/test/java/io/trino/gateway/ha/clustermonitor/TestHealthCheck.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+{
+    private Notifier notifier;
+
+    private HealthChecker healthChecker;
+
+    @BeforeEach
+    public void setUp()
+    {
+        notifier = Mockito.mock(Notifier.class);
+        healthChecker = new HealthChecker(notifier);
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/clustermonitor/TestHealthCheck.java:42]: Having these two object in fields require single threaded test. I would move to each test method.

File: gateway-ha/src/test/java/io/trino/gateway/module/TestClusterStateListenerModule.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import io.trino.gateway.ha.clustermonitor.HealthChecker;
+import io.trino.gateway.ha.clustermonitor.TrinoClusterStatsObserver;
+import io.trino.gateway.ha.config.HaGatewayConfiguration;
+import io.trino.gateway.ha.config.MonitorConfiguration;
+import io.trino.gateway.ha.config.NotifierConfiguration;
+import io.trino.gateway.ha.module.ClusterStateListenerModule;
+import io.trino.gateway.ha.router.BackendStateManager;
+import io.trino.gateway.ha.router.RoutingManager;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/module/TestClusterStateListenerModule.java:26]: Use AssertJ.

File: gateway-ha/src/main/java/io/trino/gateway/ha/notifier/EmailNotifier.java [hunk 7]
```diff
@@ -54,10 +54,11 @@ public EmailNotifier(NotifierConfiguration notifierConfiguration)
     @Override
     public void sendNotification(String subject, String content)
     {
+        content = notifierConfiguration.getCustomContent() != null ? notifierConfiguration.getCustomContent() : content;
```
[the remark this answers is not in the document]
Comment (prakhar10) [reply gateway-ha/src/main/java/io/trino/gateway/ha/notifier/EmailNotifier.java]: I tried `firstNonNull` but the maven modernizer plugin suggested to use `requireNonNullElse` instead.

File: gateway-ha/src/main/java/io/trino/gateway/ha/module/ClusterStateListenerModule.java [hunk 8]
```diff
@@ -50,10 +53,14 @@ public List<TrinoClusterStatsObserver> getClusterStatsObservers(
...
     {
-        return ImmutableList.<TrinoClusterStatsObserver>builder()
-                .add(new HealthCheckObserver(mgr))
-                .add(new ClusterStatsObserver(backendStateManager))
-                .build();
+        NotifierConfiguration notifierConfiguration = getConfiguration().getNotifier();
+        ImmutableList.Builder<TrinoClusterStatsObserver> observerBuilder = ImmutableList.builder();
+        observerBuilder.add(new HealthCheckObserver(mgr));
+        observerBuilder.add(new ClusterStatsObserver(backendStateManager));
+        if (notifierConfiguration.isEnabled()) {
```
[the remark this answers is not in the document]
Comment (prakhar10) [reply gateway-ha/src/main/java/io/trino/gateway/ha/module/ClusterStateListenerModule.java:60]: I think we cannot get `monitorConfig.getNotifier()` because monitorConfig is an object of `MonitorConfiguration` and not `NotifierConfiguration`

File: gateway-ha/src/test/java/io/trino/gateway/ha/notifier/TestEmailNotifier.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+{
+    @Test
+    public void testSendNotificationCustomContent()
+    {
+        NotifierConfiguration notifierConfiguration = mock(NotifierConfiguration.class);
+        when(notifierConfiguration.getCustomContent()).thenReturn("Custom content");
+
+        EmailNotifier emailNotifier = new EmailNotifier(notifierConfiguration);
+
+        emailNotifier.sendNotification("Test Subject", "Test Content");
```
[the remark this answers is not in the document]
Comment (prakhar10) [reply gateway-ha/src/test/java/io/trino/gateway/ha/notifier/TestEmailNotifier.java:42]: yeah it was just basically to see if there are no errors produced since the method returns void

**Comment by mosabua** [comment]
@prakhar10 can you address to the feedback comments

**Comment by prakhar10** [comment]
Branch was messed up, created a new PR - https://github.com/trinodb/trino-gateway/pull/319
